File: python/2008.py
from typing import List

class Solution:
    def maxTaxiEarnings(self, n: int, rides: List[List[int]]) -> int:
        dp = [0] * (n + 1)
        g = [[] for _ in range(n+1)]
        for x, y, tip in rides:
            g[y].append([x, tip])
        for i in range(1,n+1):
            dp[i] = dp[i-1]
            for x, tip in g[i]:
                dp[i] = max(dp[i], dp[x]+i-x+tip)
        return dp[-1]

File: go/2008.go
package main

// type P struct {
// 	from int
// 	tip  int
// }

// func maxTaxiEarnings(n int, rides [][]int) int64 {
// 	dp := make([]int64, n+1)
// 	g := make([][]P, n+1)
// 	for _, ride := range rides {
// 		x, y, p := ride[0], ride[1], ride[2]
// 		g[y] = append(g[y], P{x, p})
// 	}
// 	for i := 1; i < n+1; i++ {
// 		dp[i] = dp[i-1]
// 		for _, p := range g[i] {
// 			dp[i] = max(dp[i], dp[p.from]+int64(i-p.from+p.tip))
// 		}
// 	}
// 	return dp[n]
// }
